src/pmem.c

```diff
@@ -602,17 +653,243 @@ memmove_nodrain_normal(void *pmemdest, const void *src, size_t len)
+
+	if (len == 0)
+		return NULL;
+
+	if (src == pmemdest)
+		return pmemdest;
+
+	d = (__m128i *)pmemdest;
+	s = (__m128i *)src;
+
+	cnt = len >> CHUNK_SHIFT;
+	for (i = 0; i < cnt; i++) {
+		xmm0 = _mm_loadu_si128(s);
+		xmm1 = _mm_loadu_si128(s + 1);
+		xmm2 = _mm_loadu_si128(s + 2);
+		xmm3 = _mm_loadu_si128(s + 3);
+		xmm4 = _mm_loadu_si128(s + 4);
+		xmm5 = _mm_loadu_si128(s + 5);
+		xmm6 = _mm_loadu_si128(s + 6);
+		xmm7 = _mm_loadu_si128(s + 7);
+		s += 8;
+		/* Store double quadword with non-temporal hint */
+		_mm_stream_si128(d,	xmm0);
+		_mm_stream_si128(d + 1,	xmm1);
+		_mm_stream_si128(d + 2,	xmm2);
+		_mm_stream_si128(d + 3,	xmm3);
+		_mm_stream_si128(d + 4,	xmm4);
+		_mm_stream_si128(d + 5,	xmm5);
+		_mm_stream_si128(d + 6,	xmm6);
+		_mm_stream_si128(d + 7,	xmm7);
+		d += 8;
+	}
+
+	/* Copy the tail (<128 bytes) in 16 bytes chunks */
+	len &= CHUNK_MASK;
+	if (len != 0) {
+		cnt = len >> MOVNT_SHIFT;
+		for (i = 0; i < cnt; i++) {
+			xmm0 = _mm_loadu_si128(s);
+			_mm_stream_si128(d, xmm0);
+			s++;
+			d++;
+		}
+	}
+
+	/* Copy the last bytes (<16). First dwords then bytes */
+	len &= MOVNT_MASK;
+	if (len != 0) {
+		cnt = len >> DWORD_SHIFT;
+		int32_t	*d32 = (int32_t *)d;
+		int32_t	*s32 = (int32_t *)s;
+
+		for (i = 0; i < cnt; i++) {
+			_mm_stream_si32(d32, *s32);
+			d32++;
+			s32++;
+		}
+
+		cnt =	len & DWORD_MASK;
+		uint8_t	*d8 = (uint8_t *)d32;
+		const uint8_t	*s8 = (uint8_t *)s32;
+
+		for (i = 0; i < cnt; i++) {
+			*d8 = *s8;
+			d8++;
+			s8++;
+		}
+		pmem_flush(d8, cnt);
+	}
+	return tempdest;
+}
+
 /*
  * memmove_nodrain_movnt -- (internal) memmove to pmem without hw drain, movnt
  */
 static void *
 memmove_nodrain_movnt(void *pmemdest, const void *src, size_t len)
 {
+
+	int	i;
+	uint64_t chunk_size;
+	uint64_t num_chunks;
+	size_t cnt = 0;
+	void *dest1 = pmemdest;
+	void *src1 = (void *)src;
+	int mod = 0;
+
 	/* way too chatty for LOG level 3 */
 	LOG(15, "pmemdest %p src %p len %zu", pmemdest, src, len);
 
-	/* XXX stub version */
-	return memmove(pmemdest, src, len);
+	/*
+	 * The way this works is:
+	 * 	Calculate the non-overlapping size
+	 * 	Break the memmove into these chunk sizes
+	 *	Determine if we are copying forward/backward
+	 * This handles all cases, including overlap. Consider the following:
+	 *
+	 * 	src
+	 * high addr				dest
+	 *					high addr
+	 * ------
+	 * |	|s3-----------------------------------------------V
+	 * ------						--------
+	 * |	|s2-----------------------------|		|	|d3
+	 * ------				|-------|	--------
+	 * |	|s1---------------------------- 	|------>|	|d2
+	 * ------				|		--------
+	 *					|-------------->|	|d1
+	 * **** 						--------
+	 *
+	 * In this case we copy down, starting with s1 to d1, so that we
+	 * never overwrite the source before moving it. Any leftover
+	 * bytes are handled after moving the chunk size data.
+	 * This works similarly when src < dest.
+	 */
+	if (src > pmemdest)
+		chunk_size = util_nonoverlap_range((void *)src1, dest1, len);
```

Comment:
1) *src1* is of type "void *", so what's the point in casting it to "void *"?
2) Why not to use "src" and "pmemdest" here?